File: leetcode/0387-first-unique-character-in-a-string/firstUniqChar.ts
function firstUniqChar(s: string): number {
  const charOnlyOnceMap = new Map<string, boolean>();
  let result = -1;

  for (const char of s) {
    if (charOnlyOnceMap.has(char)) {
      charOnlyOnceMap.set(char, false);
      continue;
    }

    charOnlyOnceMap.set(char, true);
  }

  for (const [index, char] of Array.from(s).entries()) {
    if (!charOnlyOnceMap.get(char)) {
      continue;
    }

    result = index;
    break;
  }

  return result;
}

export default firstUniqChar;

File: leetcode/0387-first-unique-character-in-a-string/firstUniqChar.go
package leetcode

func firstUniqChar(s string) int {
	var result = -1
	var charOnlyOnceMap = map[rune]bool{}

	for _, char := range s {

		if _, ok := charOnlyOnceMap[char]; ok {
			charOnlyOnceMap[char] = false
			continue
		}

		charOnlyOnceMap[char] = true
	}

	for index, char := range s {
		if !charOnlyOnceMap[char] {
			continue
		}

		result = index
		break
	}

	return result
}
